File: Dynamic-Programming/213. House Robber II/houseRobberII.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var rob = function (nums) {
    return Math.max(nums[0], robHouse(nums, 0, nums.length - 2), robHouse(nums, 1, nums.length - 1));
};

function robHouse(nums, start, end) {
    let rob1 = 0;
    let rob2 = 0;

    for (let i = start; i <= end; i++) {
        let currentMax = Math.max(rob1 + nums[i], rob2);
        rob1 = rob2;
        rob2 = currentMax;
    }

    return rob2;
}

File: Dynamic-Programming/213. House Robber II/houseRobberII.py
class Solution:
    def rob(self, nums: List[int]) -> int:
        def robOption(nums, start, end):
            rob1 = 0
            rob2 = 0
            
            for i in range(start, end):
                cur_max = max(rob1 + nums[i], rob2)
                rob1 = rob2
                rob2 = cur_max
            
            return rob2
        
        return max(
            # -1 will not work as len(nums) - 1, only when getting an element nums[-1] works, which gets the last element in the array
            nums[0], robOption(nums, 0, len(nums) - 1), robOption(nums, 1, len(nums))
        )
